### ISSUE TITLE
The study with the highest priority is returned by announce, also if there are no jobs left for it

### ISSUE THREAD

**Issue by smathot** [open-cogsci/omm-server#63]
I just tested a scenario with two studies queued for a single participant. When the first experiment is finished, the announce endpoint should return the second study. But it doesn't seem to do that; that is, the announce end point always returns the study with the highest priority for a participant, even if there are no jobs left. (At least, that's my understanding of the problem based on some testing.)

**Comment by dschreij** [comment]
I have a feeling this has the same underlying problem as the participation trend lines showing up as flat even while there have been participations. Namely, the status of studies and/or jobs is not updated correctly.
The query/function that retrieves studies for `/announce` has this filter:

```javascript
const study = await ptcp.studies()
      ...
      .whereInPivot('status_id', [1, 2])
      ...
      .firstOrFail()
```
which should only return studies that are pending (status_id = 1) or in progress (status_id = 2). I think this problem occurs because bookkeeping functions that keep track of study and job states are not working as intended. A study with no jobs left should have been set to a status_id of 3 (completed). There are two potential cause I can currently think of:

- The bookkeeping functions are broken
- The bookkeeping functions are not being called, because jobs states are changed by endpoints/functions that do not call these functions. 

A study is set to complete if there are no more open jobs after results have been posted to `POST /participants/{identifier}/{job_id}/result`. This function/endpoint specifically contains the bookkeeping functions. So my first question is, do you use this endpoint to store results, or do you use an alternative method (an updateJobs endpoint for instance).

My second question. Can you confirm by checking in the database if the study with no more jobs left indeed does not have status_id 3, but 1 or 2 instead?

**Comment by dschreij** [comment]
I can reproduce this locally, and I don't even have to check the database. In Statistics panel of a study, participants who have done all jobs are still indicates as 'in progress'.

<img width="580" alt="Screenshot 2021-02-22 at 21 53 12" src="https://user-images.githubusercontent.com/689044/108768886-8db5b600-7558-11eb-86c5-e64e3f9aa32d.png">

Can you confirm this for your case as well?

**Comment by smathot** [comment]
Yes, I can confirm that. This is what I see after the participant has completed all jobs of this study.

![image](https://user-images.githubusercontent.com/594936/108827821-5b42a200-75c6-11eb-8583-2685030b692b.png)

**Comment by dschreij** [comment]
I already walked to the process with a debugger and found the section where this goes wrong. Seems like a minor problem which should be easily fixed. I should be able to do this in the next days.

**Comment by dschreij** [comment]
I fixed this in commit https://github.com/open-cogsci/omm-server/commit/65119f2f261e79302768ce5217d9852248f921a0, but accidentally attributed the commit to https://github.com/open-cogsci/omm-server/issues/67. I don't feel like starting an interactive rebase and the mess it may entail, so I'll just leave it like this. In any case, this should be fixed. Can you pull the latest changes from the develop branch and give this a spin?
I did discover another bug in the process of this fix, but I will open a separate issue for it.

**Comment by smathot** [comment]
There's an issue with the latest snapshot from `development` that prevents me from properly testing it. Namely, when I try to change the priority of a participation, I get an error: 
![image](https://user-images.githubusercontent.com/594936/108990682-7b8e6180-7697-11eb-8bdd-aadd07d5774d.png)

**Comment by dschreij** [comment]
I was baffled by this because I didn't change anything to ther priority dial, but then I took a closer look at your error message and the problem lies here:

![image](https://user-images.githubusercontent.com/689044/108999989-615a8080-76a3-11eb-9b60-85d52658bf0a.png)

the domain in your API_URL differs from the domain this call is made from..... Welcome to the wonderful world of dreadful and confusing CORS errors. I you make sure your API url also contains localhost instead of 127.0.0.1 (I know, I know, they are the same) then this should work.

**Comment by dschreij** [comment]
My suspicion was correct. See
https://github.com/open-cogsci/omm-server/blob/master/config/cors.js#L32
So either we add 'PATCH' to this list to basically allow PATCH requests from every domain, or we have to make sure that the API URL always has the same root domain as that of the app itself (which is the case for omm.cogsci.nl anyway).
CORS is basically a security measure, but I've more often seen it do harm than good 😅

**Comment by smathot** [comment]
It works! The error when updating the priority was indeed due to the use of `localhost` rather than `127.0.0.1`. And announce now automatically advances to the next experiment in line if the first one is finished! So I think this issue can be closed.

**Comment by dschreij** [comment]
That's a relief! The issue will be automatically closed once these changes are merged into master. I plan on doing this as soon as I have fixed #67 and #69 too.
